notifications/notifiers/email.py

```diff
@@ -54,22 +55,30 @@ def send_notification(self, email_notification):
         Args:
             email_notification (EmailNotification): the notification to be sent
         """
+        messages = None
         user = email_notification.user
         notification_settings = NotificationSettings.objects.get(
             user=user,
             notification_type=email_notification.notification_type,
         )
-        last_notification = self._get_most_recent_notification(notification_settings, before=email_notification)
 
-        data = self._get_notification_data(notification_settings, last_notification)
+        with utils.mark_as_sent_or_canceled(email_notification) as will_send:
+            if not will_send:
+                return
 
-        # generate the message (there's only 1)
-        messages = list(api.messages_for_recipients([
-            (recipient, user, data) for recipient, user in api.safe_format_recipients([user])
-        ], self._template_name))
+            last_notification = self._get_most_recent_notification(notification_settings, before=email_notification)
 
-        for message in messages:
-            try:
-                utils.send_at_most_once(email_notification, message)
-            except SendEmailsException:
-                log.exception("Error sending email")
+            data = self._get_notification_data(notification_settings, last_notification)
+
+            # generate the message (there's only 1)
+            messages = list(api.messages_for_recipients([
+                (recipient, user, data) for recipient, user in api.safe_format_recipients([user])
+            ], self._template_name))
+
+            if not messages:
+                raise CancelNotificationError()
```

Comment:
Maybe add a unit test to verify `CancelNotificationError` is raised in this case.